observability/micrometer/src/main/java/org/wildfly/extension/micrometer/MicrometerSubsystemRegistrar.java

```diff
@@ -0,0 +1,189 @@
+                .addCapability(MICROMETER_COLLECTOR_RUNTIME_CAPABILITY)
+                .addAttributes(List.of(ENDPOINT, STEP, EXPOSED_SUBSYSTEMS))
+                .withDeploymentChainContributor(target -> {
+                    target.addDeploymentProcessor(MicrometerConfigurationConstants.NAME, DEPENDENCIES, DEPENDENCIES_MICROMETER,
+                            new MicrometerDependencyProcessor());
+                    target.addDeploymentProcessor(MicrometerConfigurationConstants.NAME, POST_MODULE, POST_MODULE_MICROMETER,
+                            new MicrometerDeploymentProcessor(exposeAnySubsystem, exposedSubsystems, wildFlyRegistry));
+                })
+                .build();
+
+        ManagementResourceRegistrar.of(descriptor).register(registration);
+
+        return registration;
+    }
+
+    @Override
+    public ResourceServiceInstaller configure(OperationContext context, ModelNode model) throws OperationFailedException {
+        exposedSubsystems = MicrometerSubsystemRegistrar.EXPOSED_SUBSYSTEMS.unwrap(context, model);
+        exposeAnySubsystem = exposedSubsystems.remove("*");
+        endpoint = MicrometerSubsystemRegistrar.ENDPOINT.resolveModelAttribute(context, model).asStringOrNull();
+        step = MicrometerSubsystemRegistrar.STEP.resolveModelAttribute(context, model).asLong();
+        wildFlyRegistry = endpoint != null ?
+                new WildFlyOtlpRegistry(new WildFlyMicrometerConfig(endpoint, step)) :
+                new NoOpRegistry();
+
+        try {
+            new JmxMicrometerCollector(wildFlyRegistry).init();
+        } catch (IOException e) {
+            throw MICROMETER_LOGGER.failedInitializeJMXRegistrar(e);
+        }
+
+        ServiceDependency<ModelControllerClientFactory> mccf = ServiceDependency.on(CLIENT_FACTORY_CAPABILITY, ModelControllerClientFactory.class);
+        ServiceDependency<Executor> executor = ServiceDependency.on(MANAGEMENT_EXECUTOR, Executor.class);
+        ServiceDependency<ProcessStateNotifier> processStateNotifier = ServiceDependency.on(PROCESS_STATE_NOTIFIER, ProcessStateNotifier.class);
+
+        Supplier<MicrometerCollector> collectorSupplier = () ->
+                new MicrometerCollector(mccf.get().createClient(executor.get()), processStateNotifier.get(), wildFlyRegistry);
+
+        AtomicReference<MicrometerCollector> captor = new AtomicReference<>();
+
+        context.addStep((operationContext, modelNode) -> {
+            MicrometerCollector collector = captor.get();
+            // Given that this step runs in the VERIFY stage, and our collector service was started eagerly, the
+            // collector reference _should_ be non-null.
+            if (collector != null) {
+                ImmutableManagementResourceRegistration rootResourceRegistration = context.getRootResourceRegistration();
+                Resource rootResource = context.readResourceFromRoot(EMPTY_ADDRESS);
```

Comment:
Why perform a fully recursive read of the whole model if we might only need to read specific subsysems?

Reply:
I do not know. :) That code has been that way (coming from the metrics subsystem) from long before my time. I'm definitely open to improving, but I can't explain/defend it. :)